cruise-control-core/src/main/java/com/linkedin/cruisecontrol/monitor/sampling/aggregator/MetricSampleAggregator.java

```diff
@@ -0,0 +1,474 @@
+    _generation.incrementAndGet();
+  }
+
+  /**
+   * Remove the given set of entity groups from the MetricSampleAggregator.
+   *
+   * @param entityGroups the entity groups to remove from the MetricSampleAggregator.
+   */
+  public void removeEntityGroup(Set<G> entityGroups) {
+    _rawMetrics.entrySet().removeIf(entry -> entityGroups.contains(entry.getKey().group()));
+    _generation.incrementAndGet();
+  }
+
+  /**
+   * Clear the MetricSampleAggregator.
+   */
+  public void clear() {
+    _windowRollingLock.lock();
+    try {
+      _rawMetrics.clear();
+      _aggregatorState.clear();
+      _generation.incrementAndGet();
+    } finally {
+      _windowRollingLock.unlock();
+    }
+  }
+
+  // package private for testing.
+  MetricSampleAggregatorState<G, E> aggregatorState() {
+    updateAggregatorStateIfNeeded();
+    return _aggregatorState;
+  }
+
+  // both from and to window indexes are inclusive.
+  private List<Long> getWindowList(long fromWindowIndex, long toWindowIndex) {
+    _windowRollingLock.lock();
+    try {
+      if (_rawMetrics.isEmpty()) {
+        return Collections.emptyList();
+      }
+      List<Long> windows = new ArrayList<>((int) (toWindowIndex - fromWindowIndex + 1));
+      for (long i = fromWindowIndex; i <= toWindowIndex; i++) {
+        windows.add(i * _windowMs);
+      }
+      return windows;
+    } finally {
+      _windowRollingLock.unlock();
+    }
+  }
+
+  private void updateAggregatorStateIfNeeded() {
+    long currentGeneration = generation();
+    if (_aggregatorState.generation() < currentGeneration) {
+      for (long windowIdx : _aggregatorState.windowIndexesToUpdate(currentGeneration, _oldestWindowIndex, _currentWindowIndex)) {
+        _aggregatorState.updateWindowState(windowIdx, getWindowState(windowIdx, currentGeneration));
+      }
+      _aggregatorState.setGeneration(currentGeneration);
+    }
+  }
+
+  private WindowState<G, E> getWindowState(long windowIdx, long currentGeneration) {
+    WindowState<G, E> windowState = new WindowState<>(currentGeneration);
+    for (Map.Entry<E, RawMetricValues> entry : _rawMetrics.entrySet()) {
+      E entity = entry.getKey();
+      RawMetricValues rawValues = entry.getValue();
+      if (rawValues.isValidAtWindowIndex(windowIdx) && rawValues.numImputations() <= _maxAllowedImputations) {
+        windowState.addValidEntities(entity);
+      }
+    }
+    return windowState;
+  }
+
+  private boolean rollOutNewWindowIfNeeded(long index) {
+    if (_currentWindowIndex < index) {
+      _windowRollingLock.lock();
+      try {
+        if (_currentWindowIndex < index) {
+          // find out how many windows we need to reset in the raw metrics.
+          int numWindowsToRollOut = (int) (index - _currentWindowIndex);
+          // First set the oldest window index so newly coming older samples will not be added.
+          long prevOldestWindowIndex = _oldestWindowIndex;
+          // The first valid window is actually 1 instead of 0.
```

Comment:
0 is invalid because the way we calculate the window index. it is `time / windowMs + 1`, so it will never be 0 unless time is negative.
It is probably clearer to say the "first possible window index is 1 instead of 0".